### PR TITLE
Postgres support different wallet/database strategies

### PR DESCRIPTION
**Pull request by ianco**
Add support for multi-wallet databases

Adds a new global struct with functionality specific to each strategy (database per wallet or multi-wallet database)

Strategy is specified as a wallet configuration param "wallet_scheme": "MultiWalletSingleTable" (default is single wallet per database)

### CODE REVIEW

**Comment by mattraffel** [comment]
Are you ready to remove WIP status on this PR?

File: ci/indy-pool.dockerfile [hunk 1]
```diff
@@ -29,7 +29,7 @@ ARG indy_node_ver=1.9.0~dev1009
 ARG python3_indy_crypto_ver=0.4.5
 ARG indy_crypto_ver=0.4.5
 
-RUN apt-get update -y && apt-get install -y \
+RUN apt-get update -y && apt-get install -y --allow-unauthenticated \
```
Comment (mattraffel) [inline ci/indy-pool.dockerfile]: do we need the allow-unauthenticated parameter?

File: experimental/plugins/postgres_storage/src/lib.rs [hunk 2]
```diff
@@ -992,10 +1009,24 @@ fn _tag_names_from_json(json: &str) -> Result<Vec<TagName>, WalletStorageError>
 #[cfg(test)]
 mod tests {
     use super::*;
+    use std::env;
     use std::ffi::{CString, CStr};
     use std::{slice, ptr};
     use wql::storage::ENCRYPTED_KEY_LEN;
 
+    // TODO We need init() to be the first test run in order to create the wallets database
+    //#[test]
```
Comment (mattraffel) [inline experimental/plugins/postgres_storage/src/lib.rs]: is this test attribute supposed to commented out?

File: experimental/plugins/postgres_storage/src/postgres_storage.rs [hunk 3]
```diff
@@ -841,8 +1430,24 @@ impl WalletStorage for PostgresStorage {
...
+        // TODO add wallet_id limitation to search
+        let query_qualifier = unsafe {
+            SELECTED_STRATEGY.query_qualifier()
+        };
+        let wallet_id_arg = self.wallet_id.as_bytes().to_owned();
         let total_count: Option<usize> = if search_options.retrieve_total_count {
-            let (query_string, query_arguments) = query::wql_to_sql_count(&type_, query)?;
+            // TODO add an extra parameter with the SELECTED_STRATEGY.query_qualifier to AND to the query
+            //let (query_string, query_arguments) = query::wql_to_sql_count(&type_, query)?;
+            let (query_string, query_arguments) = match query_qualifier {
```
Comment (matt-raffel-kiva) [inline experimental/plugins/postgres_storage/src/postgres_storage.rs]: can we remove the commented out code?

File: experimental/plugins/postgres_storage/src/postgres_storage.rs [hunk 4]
```diff
@@ -863,7 +1468,18 @@ impl WalletStorage for PostgresStorage {
                 retrieve_type: search_options.retrieve_type,
             };
 
-            let (query_string, query_arguments) = query::wql_to_sql(&type_, query, options)?;
+            // TODO add an extra parameter with the SELECTED_STRATEGY.query_qualifier to AND to the query
+            //let (query_string, query_arguments) = query::wql_to_sql(&type_, query, options)?;
```
Comment (matt-raffel-kiva) [inline experimental/plugins/postgres_storage/src/postgres_storage.rs]: same here

File: experimental/plugins/postgres_storage/src/postgres_storage.rs [hunk 5]
```diff
@@ -883,6 +1499,7 @@ impl WalletStorage for PostgresStorage {
     fn close(&mut self) -> Result<(), WalletStorageError> {
         //let pool = self.pool.clone();
         //let conn = pool.get().unwrap();
+        // TODO throws a borrow error; temporary workaround is to rely on idle connection timeout
         //let _ret = conn.finish();
```
Comment (matt-raffel-kiva) [inline experimental/plugins/postgres_storage/src/postgres_storage.rs]: same here, please :)

File: ci/indy-pool.dockerfile (hunk 1 above)
Comment (jovfer) [reply ci/indy-pool.dockerfile]: SF's keys were updated then `indy-pool.dockerfile` in master was updated. So allow-unauthenticated can be removed from the PR after merge with master

File: experimental/plugins/postgres_storage/src/lib.rs (hunk 2 above)
Comment (ianco) [reply experimental/plugins/postgres_storage/src/lib.rs]: Yes this test is not necessary, storage initialization is called on each test

**Comment by ianco** [comment]
Testing is still in progress, should be ready to remove the WIP status tonight (Wed)

**Comment by ianco** [comment]
Fixed some issues with the wallet search, and added some updated docs.

**Pull request merged**